Source/WebCore/Modules/url-pattern/URLPatternParser.cpp

```diff
@@ -0,0 +1,341 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternParser.h"
+
+#include "URLPatternTokenizer.h"
+
+namespace WebCore {
+namespace URLPatternUtilities {
+
+URLPatternParser::URLPatternParser(EncodingCallbackType urlComponentType, String&& segmentWildcardRegexp)
+    : m_callbackType(urlComponentType)
+    , m_segmentWildcardRegexp(segmentWildcardRegexp)
+{
+}
+
+ExceptionOr<void> URLPatternParser::performParse(const URLPatternStringOptions& patternSettings)
+{
+    ExceptionOr<void> maybeFunctionException;
+
+    while (m_index < m_tokenList.size()) {
+        auto charToken = tryToConsumeToken(TokenType::Char);
+        auto nameToken = tryToConsumeToken(TokenType::Name);
+        auto regexOrWildcardToken =  tryToConsumeRegexOrWildcardToken(nameToken);
+
+        if (nameToken.isInitialized() || regexOrWildcardToken.isInitialized()) {
+            String prefix = emptyString();
+
+            if (charToken.isInitialized())
+                prefix = charToken.value.toString();
+
+            if (!prefix.isEmpty() && prefix != patternSettings.prefixCodepoint) {
+                m_pendingFixedValue = makeString(WTFMove(m_pendingFixedValue), prefix);
+                prefix = emptyString();
+            }
+
+            maybeFunctionException = maybeAddPartFromPendingFixedValue();
+            if (maybeFunctionException.hasException())
+                return maybeFunctionException.releaseException();
+
+            auto modifierToken = tryToConsumeModifierToken();
+
+            maybeFunctionException = addPart(WTFMove(prefix), nameToken, regexOrWildcardToken, ""_s, modifierToken);
```

Comment:
Why use `""_s` instead of `emptyString()`?

Reply:
I was implementing this part of the spec where suffix is the empty string. 
<img width="909" alt="Screenshot 2024-11-01 at 10 42 59 AM" src="https://github.com/user-attachments/assets/869f5c93-a509-4951-831f-9374eb22ad21">

Since suffix is not re-used at all, I wanted to transfer ownership of the suffix string to the part list. You can see in other parts, I do WTFMove(suffix). 

But emptyString() is const so I couldn't transfer ownership. So I used ""_s instead.

Reply:
So use `const String&` instead of `String&&` for the parameter? As addressed elsewhere in my review feedback, you're passing a `String&&` around and end up calling `callEncodingCallback()` which doesn't make use of the r-value reference anyway (no internal WTFMove()).